src/panoptes/pocs/focuser/birger.py

```diff
@@ -249,36 +253,32 @@ def _send_command(self, command, response_length=None):
         # Clear the input buffer in case there's anything left over in there.
         self._serial_port.reset_input_buffer()
 
-        # Send command
-        self._serial_io.write(command + '\r')
-
-        # In verbose mode adaptor will first echo the command
-        echo = self._serial_io.readline().rstrip()
-        assert echo == command, self.logger.warning("echo != command: {} != {}".format(
-            echo, command))
-
-        # Adaptor should then send 'OK', even if there was an error.
-        ok = self._serial_io.readline().rstrip()
-        assert ok == 'OK'
-
         # Depending on which command was sent there may or may not be any further
         # response.
         response = []
 
-        if response_length == 0:
-            # Not expecting any further response. Should check the buffer anyway in case an error
-            # message has been sent.
-            if self._serial_port.in_waiting:
-                response.append(self._serial_io.readline())
+        for i in range(self._max_command_attempts):
+            # Send the command
+            self._serial_io.write(command + '\r')
 
-        elif response_length > 0:
-            # Expecting some number of lines of response. Attempt to read that many lines.
-            for i in range(response_length):
-                response.append(self._serial_io.readline())
+            # In verbose mode adaptor will first echo the command
+            echo = self._serial_io.readline().rstrip()
 
-        else:
-            # Don't know what to expect. Call readlines() to get whatever is there.
-            response.append(self._serial_io.readlines())
+            # Adaptor should then send 'OK', even if there was an error.
+            ok = self._serial_io.readline().rstrip()
+
+            if echo == command and ok == 'OK':
+                # Get any avaialble response left in buffer.
+                response.append(self._serial_io.readlines())
```

Comment:
You've removed some of the functionality of the original code here. Almost all of the Birger commands have an expected response length, i.e. how many lines of response there should be after the echo and the 'OK'. 

`self._serial_io.readlines()` is supposed to be the fall back option only used for the 1 or 2 commands that return an unknown number of lines of response because it will block for the duration of the serial timeout waiting for more lines of response. When the number of lines is known it is far quicker to read only the expected number of lines.

We always try to read at least 1 line though, because it's possible there will be an error message immediately after the 'OK'.

```suggestion
                if response_length == 0:
                    # Not expecting any further response. Should check the buffer anyway in case an error
￼                   # message has been sent.
                    response.append(self._serial_io.readline())
                elif response_length > 0:
                    # Expecting some number of lines of response. Attempt to read that many lines.
                    for i in range(response_length):
￼                        response.append(self._serial_io.readline())
                else:
                    # Don't know what to expect. Call readlines() to get whatever is there.
                    response.append(self._serial_io.readlines())
```